File: 36-valid-sudoku/valid-sudoku.cpp
#define Maximum_Digits 10
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        for(int i=0;i<9;i++){
            for(int j=0;j<9;j++){
                cout<<board[i][j]<<" ";
            }
            cout<<endl;
        }
        cout<<endl;

        // int i = 4;
        // char c = '0' + i;
        // cout<<c<<endl;

        vector<vector<int>> row(12, vector<int> (12, 0));
        vector<vector<int>> col(12, vector<int> (12, 0));
        vector<vector<int>> box(12, vector<int> (12, 0));
        vector<pair<int,int>>vp;
        
        for(int i=0;i<9;i++){
            for(int j=0;j<9;j++){
                if(board[i][j]=='.'){
                    vp.push_back({i,j});
                    continue;
                }

                char num = board[i][j];
                int n = num -'0';
                row[i][n]++;
                col[j][n]++;
                //if(board[i][j]=='5'){cout<<i<<" "<<j<<" "<<n<<endl;}

                int boxi,boxj;
                if(i<=2){boxi=0;}
                else if(i<=5){boxi=3;}
                else if(i<=8){boxi=6;}

                if(j<=2){boxi+=0;}
                else if(j<=5){boxi+=1;}
                else if(j<=8){boxi+=2;}

                box[boxi][n]++;

                if(row[i][n]>1 || col[j][n]>1 || box[boxi][n]>1){return false;}
            }
        }

        // for(int i=0;i<vp.size();i++){
        //     int r = vp[i].first, c = vp[i].second;
        //     int count=0;
        //     for(int j=1;j<=9;j++){

        //         int boxi,boxj;
        //         if(r<=2){boxi=0;}
        //         else if(r<=5){boxi=3;}
        //         else if(r<=8){boxi=6;}

        //         if(c<=2){boxi+=0;}
        //         else if(c<=5){boxi+=1;}
        //         else if(c<=8){boxi+=2;}

        //         if(row[r][j]==0 && col[c][j]==0 && box[boxi][j]==0){
        //             //board[r][c] = '0'+j;
        //             //cout<<r<<" "<<c<<" "<<j<<endl;
        //             row[r][j]=1 , col[c][j]=1 , box[boxi][j]=1;count=1;break;
        //         }
        //     }
        //     if(count==0){
        //         //cout<<r<<" "<<c<<endl;
        //         return false;
        //     }
        // }
        // cout<<11111111<<endl;
        // for(int i=0;i<9;i++){
        //     for(int j=0;j<9;j++){
        //         cout<<board[i][j]<<" ";
        //     }
        //     cout<<endl;
        // }
        // cout<<endl;
        // for(int i=0;i<10;i++){
        //     for(int j=0;j<10;j++){
        //         cout<<col[i][j]<<" ";
        //     }
        //     cout<<endl;
        // }
        // cout<<endl;
        // for(int i=0;i<10;i++){
        //     for(int j=0;j<10;j++){
        //         cout<<box[i][j]<<" ";
        //     }
        //     cout<<endl;
        // }
        return true;
    }
};

File: 36-valid-sudoku/valid-sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        ans = True
        for i in range(0,9):
            mySet = set()
            mySet2 = set()
            for j in range(0,9):
                # print(i, j)
                # print(j, 0)
                if board[i][j] in mySet:
                    # print('in row', i, j, board[i][j])
                    ans = False
                    break
                if board[j][i] in mySet2:
                    # print('in col', j, i, board[j][i])
                    ans = False
                    break

                if board[i][j]!='.':
                    mySet.add(board[i][j])
                if board[j][i]!='.':
                    # print(j, i)
                    mySet2.add(board[j][i])

            if ans==False:
                break
            # mySet.clear()
            # mySet2.clear()
        

        if ans != True:
            return ans
        
        i = 0
        while(i<9):
            j = 0
            
            while(j<9):
                mySet = set()
                print (i, j, ans)

                # first row
                if board[i][j] not in mySet:
                    if board[i][j]!='.': mySet.add(board[i][j])
                else: ans=False

                if board[i][j+1] not in mySet:
                    if board[i][j+1]!='.': mySet.add(board[i][j+1])
                else: ans=False

                if board[i][j+2] not in mySet:
                    if board[i][j+2]!='.': mySet.add(board[i][j+2])
                else: ans=False

                # second row
                if board[i+1][j] not in mySet:
                    if board[i+1][j]!='.': mySet.add(board[i+1][j])
                else: ans=False

                if board[i+1][j+1] not in mySet:
                    if board[i+1][j+1]!='.': mySet.add(board[i+1][j+1])
                else: ans=False

                if board[i+1][j+2] not in mySet:
                    if board[i+1][j+2]!='.': mySet.add(board[i+1][j+2])
                else: ans=False

                # second row
                if board[i+2][j] not in mySet:
                    if board[i+2][j]!='.': mySet.add(board[i+2][j])
                else: ans=False

                if board[i+2][j+1] not in mySet:
                    if board[i+2][j+1]!='.': mySet.add(board[i+2][j+1])
                else: ans=False

                if board[i+2][j+2] not in mySet:
                    if board[i+2][j+2]!='.': mySet.add(board[i+2][j+2])
                else: ans=False


                

                if ans == False:break
                j+=3
            if ans == False:break
            i+=3

        return ans
        